rmw_connextdds_common/src/common/rmw_impl.cpp

```diff
@@ -1215,21 +1215,96 @@ RMW_Connext_Subscriber::create(
     });
 
   DDS_TopicDescription * sub_topic = DDS_Topic_as_topicdescription(topic);
-
-  if (nullptr != cft_name) {
-    rmw_ret_t cft_rc =
-      rmw_connextdds_create_contentfilteredtopic(
-      ctx, dp, topic, cft_name, cft_filter, &cft_topic);
+  std::string cft_topic_name;
+  if (nullptr != cft_name || nullptr != subscriber_options->filter_expression) {
+    rmw_ret_t cft_rc = RMW_RET_OK;
+
+    if (nullptr != cft_name) {
+      cft_rc =
+        rmw_connextdds_create_contentfilteredtopic(
+        ctx, dp, topic, cft_name, cft_filter, NULL, &cft_topic);
+    } else if (nullptr != subscriber_options->filter_expression) {
+      cft_topic_name =
+        fqtopic_name + "_ContentFilterTopic" + RMW_Connext_Subscriber::get_atomic_id();
+      cft_rc =
+        rmw_connextdds_create_contentfilteredtopic(
+        ctx, dp, topic, cft_topic_name.c_str(),
+        subscriber_options->filter_expression,
+        subscriber_options->expression_parameters, &cft_topic);
+    }
 
     if (RMW_RET_OK != cft_rc) {
       if (RMW_RET_UNSUPPORTED != cft_rc) {
         return nullptr;
       }
     } else {
-      sub_topic = cft_topic;
+      if (nullptr != cft_topic) {
+        sub_topic = cft_topic;
+      }
     }
   }
 
+  DDS_DataReader * dds_reader = initialize_datareader(
+    ctx,
+    dp,
+    sub,
+    fqtopic_name,
+    qos_policies,
+    type_support,
+    subscriber_options,
+    internal,
+    sub_topic);
+
+  auto scope_exit_dds_reader_delete =
+    rcpputils::make_scope_exit(
+    [sub, dds_reader]()
+    {
+      if (DDS_RETCODE_OK !=
+      DDS_Subscriber_delete_datareader(sub, dds_reader))
+      {
+        RMW_CONNEXT_LOG_ERROR_SET(
+          "failed to delete DDS DataWriter")
+      }
+    });
+
+  RMW_Connext_Subscriber * rmw_sub_impl =
+    new (std::nothrow) RMW_Connext_Subscriber(
+    ctx,
+    dds_reader,
+    topic,
+    type_support,
+    subscriber_options->ignore_local_publications,
+    topic_created,
+    cft_topic,
+    internal);
+
+  if (nullptr == rmw_sub_impl) {
+    RMW_CONNEXT_LOG_ERROR_SET("failed to allocate RMW subscriber")
+    return nullptr;
+  }
+  rmw_sub_impl->fqtopic_name = fqtopic_name;
```

Comment:
There is no need to cache the "fully qualified" topic name, since that is already available from the `DDS_Topic` object.

When needed, you can access the value as a C string with:

```cpp
const char * const fqtopic_str = DDS_TopicDescription_get_name(DDS_Topic_as_topicdescription(this->dds_topic));
```

If needed, you can construct an `std::string` from that, and even if that ends up doing some memory alloc/copy, this shouldn't be a problem since AFAICT `fqtopic_name` in only accessed inside `MW_Connext_Subscriber::set_cft_expression_parameters()` which isn't a "critical path" function anyway, since it already does (potentially) plenty of allocation/deallocation (particularty when accessing that field).